easybuild/easyblocks/o/openmpi.py

```diff
@@ -59,41 +61,69 @@ def config_opt_unused(key, enable_opt=False):
         ]
 
         for key in config_opt_names:
-            if config_opt_unused(key, enable_opt=True):
+            if not config_opt_used(key, enable_opt=True):
                 self.cfg.update('configopts', '--enable-%s' % key)
 
+        # List of EasyBuild dependencies for which OMPI has known options
+        known_dependencies = ('CUDA', 'hwloc', 'libevent', 'libfabric', 'PMIx', 'UCX')
+        # Value to use for `--with-<dep>=<value>` if the dependency is not specified in the easyconfig
+        # No entry is interpreted as no option added at all
+        # This is to make builds reproducible even when the system libraries are changed and avoids failures
+        # due to e.g. finding only PMIx but not libevent on the system
+        unused_dep_value = dict()
+        # Known options since version 3.0 (no earlier ones checked)
+        if LooseVersion(self.version) >= LooseVersion('3.0'):
+            # Default to disable the option with "no"
+            unused_dep_value = {dep: 'no' for dep in known_dependencies}
+            # For those the default is to use an internal copy and not using any is not supported
+            for dep in ('hwloc', 'libevent', 'PMIx'):
+                unused_dep_value[dep] = 'internal'
+
         # handle dependencies
-        for dep in ['CUDA', 'hwloc', 'libevent', 'libfabric', 'PMIx', 'UCX']:
-            if config_opt_unused(dep.lower()):
-                dep_root = get_software_root(dep)
-                if dep_root:
-                    if dep == 'libfabric':
-                        opt_name = 'ofi'
-                    else:
-                        opt_name = dep.lower()
-                    self.cfg.update('configopts', '--with-%s=%s' % (opt_name, dep_root))
+        for dep in known_dependencies:
+            opt_name = dep.lower()
+            # If the option is already used, don't add it
+            if config_opt_used(opt_name):
+                continue
+
+            # libfabric option renamed in OpenMPI 3.1.0 to ofi
+            if dep == 'libfabric' and LooseVersion(self.version) >= LooseVersion('3.1'):
+                opt_name = 'ofi'
```

Comment:
Shouldn't this be moved above the "if config_opt_used" above, then we only get one such call.

Reply:
No it must not. `--with-ofi` and `--with-libfabric` are synonyms (almost), see 1 line below

Reply:
yes, but once we're at openmpi >= 3.1.0 there is only --with(out)-ofi right? So anyone using --with-libfabric for 3.1.0 or later is not going to get what the expect anyway.

Or is there some other combo  of options you're thinking of here that I don't grasp?

Reply:
No, in 3.1 --with-ofi was introduced and --with-libfabric was made a (deprecated) synonym for --with-ofi
So in <3.1 there is only --with-libfabric in in 3.1+ there are both. Hence explicitely checking both here

Reply:
Aha, that explains it. And no i didn't bother to check the source :-)

Reply:
I improved the comment ;)